<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>The LEGIT model</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h2>The LEGIT model</h2>

<p>The Latent Environmental &amp; Genetic InTeraction (LEGIT) model is an interaction model with two latent variables: \(\mathbf{g}\) a weighted sum of genetic variants (genetic score) and \(\mathbf{e}\) a weighted sum of environmental variables (environmental score). </p>

<p>Assuming \(\mathbf{g}_1\),&hellip;,\(\mathbf{g}_k\) are \(k\) genetic variants of interest, generally represented as the number of dominant alleles (0, 1, 2) or the absence or presence of a dominant allele (0, 1) and \(\mathbf{e}_1\),&hellip;,\(\mathbf{e}_s\) are \(s\) environments of interest, generally represented as ordinal scores (0, 1, 2 &hellip;) or continuous variables. We define the genetic score \(\mathbf{g}\) and environmental score \(\mathbf{e}\) as: </p>

<p>\[\mathbf{g}=\sum_{i=1}^{k} p_i \mathbf{g}_j \] 
\[\mathbf{e}=\sum_{l=1}^{s} q_l \mathbf{e}_l \]
where \(\|p\|_1=\sum_{i=1}^{k}p_j=1\) and \(\|q\|_1= \sum_{l=1}^{s}q_l=1\). The weights can be interpreted as the relative contributions of each genetic variant or environment.</p>

<p>The weights of these scores are estimated within a generalized linear model of the form :</p>

<p>\[\mathbf{E}[\mathbf{y}] = g^{-1}(f(\mathbf{g},\mathbf{e},X|\beta)+X_{covs} \beta_{covs})\]
where \(g\) is the link function, \(f\) is a linear function between \(\mathbf{g}\), \(\mathbf{e}\) and other variables from the matrix \(X\) and \(X_{covs}\) is a matrix of additionnal covariates.</p>

<p>For a two-way \(G\times E\) model, we would define \(f\) as:</p>

<p>\[f(\mathbf{g},\mathbf{e},X|\beta) = \beta_0+\beta_e \mathbf{e}+\beta_g \mathbf{g}+\beta_{eg} \mathbf{e}\mathbf{g}\]</p>

<p>For a three-way \(G\times E \times z\) model, where \(z\) is a known variable, we would define \(f\) as:</p>

<p>\[f(\mathbf{g},\mathbf{e},X|\beta) = \beta_0+\beta_e \mathbf{e}+\beta_g \mathbf{g}+\beta_z \mathbf{z}+\beta_{eg} \mathbf{eg}+\beta_{ez} \mathbf{ez}+\beta_{zg} \mathbf{zg}+\beta_{egz} \mathbf{egz}\]</p>

<h2>Algorithm</h2>

<p>Full details of the algorithm are available on <a href="arxiv.org/abs/1703.08111">arXiv</a>.</p>

<h2>Implementation</h2>

<p>In the LEGIT package, we include the following main functions:</p>

<ul>
<li>LEGIT: Fits a LEGIT model</li>
<li>LEGIT_cv: Uses cross-validation on a LEGIT model</li>
<li>stepwise_search: Adds the best variable or drops the worst variable one at a time in the genetic or environmental score of a LEGIT model</li>
</ul>

<h2>Notes</h2>

<ul>
<li>Interactions inside the genetic and environmental scores must be manually coded as new variables (ex: g1, g2, g1_g2 where g1_g2=g1*g2).</li>
<li>Default starting point is set to \(1/k\) for each of the genetic variants and \(1/s\) for each of the environmental variables.</li>
<li>Only local convergence is guaranteed, therefore it is recommended to try different starting points. </li>
<li>Experimentally, we rarely ended up stuck in a local optimum but when we did it was when using a starting point which assumed that every genetic variant had a positive weight while some genetic variants had positive weights and some genetic variants had negative weights. Adding genetic variants in a stepwise manner while readjusting the starting point at every step is thus recommended (This is done automatically in the stepwise_search function).</li>
</ul>

<h2>Example 1</h2>

<p>Let&#39;s look at a three-way interaction model with continuous outcome:</p>

<p>\[\mathbf{g}_j \sim Binomial(n=1,p=.30) \ j = 1, 2, 3, 4\]
\[\mathbf{e}_l \sim Normal(\mu=0,\sigma=1.5) \ l = 1, 2, 3\]
\[\mathbf{g} = .2\mathbf{g}_1 + .15\mathbf{g}_2 - .3\mathbf{g}_3 + .1\mathbf{g}_4 + .05\mathbf{g}_1\mathbf{g}_3 + .2\mathbf{g}_2\mathbf{g}_3 \] 
\[ \mathbf{e} = -.45\mathbf{e}_1 + .35\mathbf{e}_2 + .2\mathbf{e}_3\]
\[\mathbf{\mu} = -2 + 2\mathbf{g} + 3\mathbf{e} + \mathbf{z} + 5\mathbf{ge} + 2\mathbf{gz} - 1.5\mathbf{ez} + 2\mathbf{gez} \] 
\[ y \sim Normal(\mu,\sigma=1)\]</p>

<p>Let&#39;s load the package and look at the dataset.</p>

<pre><code class="r">library(LEGIT)
</code></pre>

<pre><code>## Loading required package: formula.tools
</code></pre>

<pre><code>## Loading required package: operator.tools
</code></pre>

<pre><code class="r">example_3way(N=5, sigma=1, logit=FALSE, seed=7)
</code></pre>

<pre><code>## $data
##             y     y_true        z
## 1  3.80723699 4.67808801 3.184193
## 2  7.96819988 7.24948932 3.752280
## 3  4.49051121 4.37985833 3.591745
## 4 -0.01796159 0.06050518 2.016947
## 5  0.71520307 1.13569353 2.723936
## 
## $G
##   g1 g2 g3 g4 g1_g3 g2_g3
## 1  1  1  0  0     0     0
## 2  0  0  0  0     0     0
## 3  0  1  1  0     0     1
## 4  0  0  0  1     0     0
## 5  0  0  0  0     0     0
## 
## $E
##          e1           e2        e3
## 1 0.5354793  0.701520767  1.259626
## 2 4.0751277 -1.340701085  1.058013
## 3 3.4221779 -0.460992449  1.958947
## 4 0.4860308 -0.007233633 -2.081994
## 5 2.8441006  1.482246224  1.909375
## 
## $coef_G
## [1]  0.20  0.15 -0.30  0.10  0.05  0.20
## 
## $coef_E
## [1] -0.45  0.35  0.20
## 
## $coef_main
## [1] 5.0 2.0 3.0 1.0 5.0 1.5 2.0 2.0
</code></pre>

<p>Currently &ldquo;data&rdquo; contains the outcome, the true outcome without the noise and the covariate \(z\). This dataset should always contain the outcome and all covariates (except for the genes and environments from \(\mathbf{g}\) and \(\mathbf{e}\)).&ldquo;G&rdquo; contains the genetic variants and &ldquo;E&rdquo; contains the environments. This is all you need to fit a LEGIT model.</p>

<p>We generate N=250 training observations and 100 test observations.</p>

<pre><code class="r">train = example_3way(N=250, sigma=1, logit=FALSE, seed=7)
test = example_3way(N=100, sigma=1, logit=FALSE, seed=6)
</code></pre>

<p>We fit the model with the default starting point.</p>

<pre><code class="r">fit_default = LEGIT(train$data, train$G, train$E, y ~ G*E*z)
</code></pre>

<pre><code>## Converged in 11 iterations
</code></pre>

<pre><code class="r">summary(fit_default)
</code></pre>

<pre><code>## $fit_main
## 
## Call:
## stats::glm(formula = formula, family = family, data = data, model = FALSE, 
##     y = FALSE)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -2.94965  -0.62453   0.08363   0.62145   2.65154  
## 
## Coefficients: (-7 not defined because of singularities)
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) -1.77075    0.22342  -7.926 9.09e-14 ***
## G            0.15934    1.12858   0.141    0.888    
## E           -2.86208    0.27217 -10.516  &lt; 2e-16 ***
## z            0.94193    0.06704  14.050  &lt; 2e-16 ***
## G:E         -5.69624    1.38253  -4.120 5.25e-05 ***
## G:z          2.61973    0.34190   7.662 4.77e-13 ***
## E:z          1.43411    0.07991  17.947  &lt; 2e-16 ***
## G:E:z       -1.82548    0.41205  -4.430 1.44e-05 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for gaussian family taken to be 1.065046)
## 
##     Null deviance: 2414.03  on 249  degrees of freedom
## Residual deviance:  250.29  on 235  degrees of freedom
## AIC: 741.75
## 
## Number of Fisher Scoring iterations: 2
## 
## 
## $fit_genes
## 
## Call:
## stats::glm(formula = formula_b, family = family, data = data, 
##     model = FALSE, y = FALSE)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -2.95636  -0.63173   0.07961   0.62078   2.65154  
## 
## Coefficients: (-9 not defined because of singularities)
##       Estimate Std. Error t value Pr(&gt;|t|)    
## g1     0.18333    0.01051  17.447  &lt; 2e-16 ***
## g2     0.15867    0.01422  11.156  &lt; 2e-16 ***
## g3    -0.29336    0.01269 -23.115  &lt; 2e-16 ***
## g4     0.09770    0.01090   8.966  &lt; 2e-16 ***
## g1_g3  0.06273    0.02247   2.791  0.00568 ** 
## g2_g3  0.20421    0.02329   8.770 3.68e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for gaussian family taken to be 1.06503)
## 
##     Null deviance: 2414.03  on 250  degrees of freedom
## Residual deviance:  250.28  on 235  degrees of freedom
## AIC: 741.75
## 
## Number of Fisher Scoring iterations: 2
## 
## 
## $fit_env
## 
## Call:
## stats::glm(formula = formula_c, family = family, data = data, 
##     model = FALSE, y = FALSE)
## 
## Deviance Residuals: 
##      Min        1Q    Median        3Q       Max  
## -2.94953  -0.62434   0.08404   0.62125   2.65128  
## 
## Coefficients: (-12 not defined because of singularities)
##    Estimate Std. Error t value Pr(&gt;|t|)    
## e1  0.43440    0.01486   29.24   &lt;2e-16 ***
## e2 -0.35307    0.01756  -20.11   &lt;2e-16 ***
## e3 -0.21253    0.01614  -13.17   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for gaussian family taken to be 1.065045)
## 
##     Null deviance: 2414.03  on 250  degrees of freedom
## Residual deviance:  250.29  on 235  degrees of freedom
## AIC: 741.75
## 
## Number of Fisher Scoring iterations: 2
</code></pre>

<p>This is very close to the original model. We can now use the test dataset to find the validation \(R^2\).</p>

<pre><code class="r">ssres = sum((test$data$y - predict(fit_default, test$data, test$G, test$E))^2)
sstotal = sum((test$data$y - mean(test$data$y))^2)
R2 = 1 - ssres/sstotal
R2
</code></pre>

<pre><code>## [1] 0.8809351
</code></pre>

<p>Now, let&#39;s see what happens if we introduce variables that are not in the model. Let&#39;s add these irrelevant genetic variants:</p>

<p>\[\mathbf{g'}_b \sim Binomial(n=1,p=.30) \ b = 1, 2, 3, 4, 5\]</p>

<pre><code class="r">g1_bad = rbinom(250,1,.30)
g2_bad = rbinom(250,1,.30)
g3_bad = rbinom(250,1,.30)
g4_bad = rbinom(250,1,.30)
g5_bad = rbinom(250,1,.30)
train$G = cbind(train$G, g1_bad, g2_bad, g3_bad, g4_bad, g5_bad)
</code></pre>

<p>Let&#39;s do a forward search of the genetic variants using the BIC and see if we can recover the right subset of variables.</p>

<pre><code class="r">forward_genes_BIC = stepwise_search(train$data, genes_extra=train$G, env_original=train$E, formula=y ~ E*G*z, search_type=&quot;forward&quot;, search=&quot;genes&quot;, search_criterion=&quot;BIC&quot;, interactive_mode=FALSE)
</code></pre>

<pre><code>## Keeping only variables with p-values smaller than 0.2 and which inclusion decrease the AIC
## Forward search of the genes to find the model with the lowest BIC
## 
## [Iteration: 1]
## Adding gene: g3 (Criterion before = Inf; after= 1121.89913)
## [Iteration: 2]
## Adding gene: g2 (Criterion before = 1121.89913; after= 1039.22899)
## [Iteration: 3]
## Adding gene: g1 (Criterion before = 1039.22899; after= 884.9214)
## [Iteration: 4]
## Adding gene: g4 (Criterion before = 884.9214; after= 842.92708)
## [Iteration: 5]
## Adding gene: g2_g3 (Criterion before = 842.92708; after= 796.03689)
## [Iteration: 6]
## Adding gene: g1_g3 (Criterion before = 796.03689; after= 795.10485)
## [Iteration: 7]
## No gene added
</code></pre>

<p>We recovered the right subset! Now what if we did a backward search using the AIC?</p>

<pre><code class="r">backward_genes_AIC = stepwise_search(train$data, genes_original=train$G, env_original=train$E, formula=y ~ E*G*z, search_type=&quot;backward&quot;, search=&quot;genes&quot;, search_criterion=&quot;AIC&quot;, interactive_mode=FALSE)
</code></pre>

<pre><code>## Dropping only variables with p-values bigger than 0.01 and which removal decrease the AIC
## Backward search of the genes to find the model with the lowest AIC
## 
## [Iteration: 1]
## Removing gene: g3_bad (Criterion before = 749.77592; after= 747.50098)
## [Iteration: 2]
## Removing gene: g5_bad (Criterion before = 747.50098; after= 745.40215)
## [Iteration: 3]
## Removing gene: g4_bad (Criterion before = 745.40215; after= 743.57487)
## [Iteration: 4]
## Removing gene: g1_bad (Criterion before = 743.57487; after= 741.8472)
## [Iteration: 5]
## Removing gene: g2_bad (Criterion before = 741.8472; after= 741.75015)
## [Iteration: 6]
## No gene removed
</code></pre>

<p>We deleted the irrevelant genes and obtained the right subset of variables! The stepwise_search function also has an interactive mode where the user decides which variable should be added/dropped at every step. We can only show the first iteration because the algorithm does&#39;nt receive an input from the user in the vignette but normally you can control the variables added or removed from the stepwise search. This is what the interactive mode looks like:</p>

<pre><code class="r">forward_genes_BIC = stepwise_search(train$data, genes_extra=train$G, env_original=train$E, formula=y ~ E*G*z, search_type=&quot;bidirectional-forward&quot;, search=&quot;genes&quot;, search_criterion=&quot;BIC&quot;, interactive_mode=TRUE)
</code></pre>

<pre><code>## &lt;&lt;~ Interative mode enabled ~&gt;&gt;
## Keeping only variables with p-values smaller than 0.2 and which inclusion decrease the AIC
## Dropping only variables with p-values bigger than 0.01 and which removal decrease the AIC
## Bidirectional search of the genes to find the model with the lowest BIC
## 
## [Iteration: 1]
##   variable N_old N_new p_value AIC_old  AIC_new BIC_old  BIC_new cv_R2_old
## 1       g3    NA   250 0.00000     Inf 1086.685     Inf 1121.899        NA
## 2       g1    NA   250 0.00000     Inf 1093.552     Inf 1128.766        NA
## 3       g2    NA   250 0.00000     Inf 1135.100     Inf 1170.314        NA
## 4   g1_bad    NA   250 0.00000     Inf 1155.719     Inf 1190.933        NA
## 5       g4    NA   250 0.00013     Inf 1164.285     Inf 1199.500        NA
##   cv_R2_new cv_AUC_old cv_AUC_new
## 1        NA         NA         NA
## 2        NA         NA         NA
## 3        NA         NA         NA
## 4        NA         NA         NA
## 5        NA         NA         NA
## Enter the index of the variable to be added: 
## No gene added
</code></pre>

<p>Manually forcing \(\mathbf{g}_3\) inclusion since the interactive mode cannot progress without a user, we get that the second iteration is:</p>

<pre><code class="r">forward_genes_BIC = stepwise_search(train$data, genes_original=train$G[,3,drop=FALSE], genes_extra=train$G[,-3], env_original=train$E, formula=y ~ E*G*z, search_type=&quot;bidirectional-forward&quot;, search=&quot;genes&quot;, search_criterion=&quot;BIC&quot;, interactive_mode=TRUE)
</code></pre>

<pre><code>## &lt;&lt;~ Interative mode enabled ~&gt;&gt;
## Keeping only variables with p-values smaller than 0.2 and which inclusion decrease the AIC
## Dropping only variables with p-values bigger than 0.01 and which removal decrease the AIC
## Bidirectional search of the genes to find the model with the lowest BIC
## 
## [Iteration: 1]
##   variable N_old N_new  p_value  AIC_old  AIC_new  BIC_old  BIC_new
## 1       g2   250   250 0.000000 1086.685 1000.493 1121.899 1039.229
## 2    g2_g3   250   250 0.000000 1086.685 1006.253 1121.899 1044.989
## 3       g1   250   250 0.000000 1086.685 1013.285 1121.899 1052.021
## 4    g1_g3   250   250 0.000002 1086.685 1068.051 1121.899 1106.787
## 5       g4   250   250 0.012411 1086.685 1084.133 1121.899 1122.869
##   cv_R2_old cv_R2_new cv_AUC_old cv_AUC_new
## 1        NA        NA         NA         NA
## 2        NA        NA         NA         NA
## 3        NA        NA         NA         NA
## 4        NA        NA         NA         NA
## 5        NA        NA         NA         NA
## Enter the index of the variable to be added: 
## No gene added
</code></pre>

<p>With the interactive mode, you can try alternative pathways, rather than simply adding/dropping the best/worst variable everytime.</p>

<h2>Example 2</h2>

<p>Let&#39;s take a quick look at a simple two-way example with binary outcome:</p>

<p>\[\mathbf{g}_j \sim Binomial(n=1,p=.30) \ j = 1, 2, 3, 4\]
\[\mathbf{e}_l \sim Normal(\mu=0,\sigma=1.5) \ l = 1, 2, 3\]
\[\mathbf{g} = .2\mathbf{g}_1 + .15\mathbf{g}_2 - .3\mathbf{g}_3 + .1\mathbf{g}_4 + .05\mathbf{g}_1\mathbf{g}_3 + .2\mathbf{g}_2\mathbf{g}_3 \] 
\[ \mathbf{e} = -.45\mathbf{e}_1 + .35\mathbf{e}_2 + .2\mathbf{e}_3\]
\[\mathbf{\mu} = -1 + 2\mathbf{g} + 3\mathbf{e} + 4\mathbf{ge} \]
\[ y \sim Binomial(n=1,p=logit(\mu))\]</p>

<p>We generate N=1000 training observations.</p>

<pre><code class="r">library(LEGIT)
train = example_2way(N=1000, logit=TRUE, seed=777)
</code></pre>

<p>We fit the model with the default starting point.</p>

<pre><code class="r">fit_default = LEGIT(train$data, train$G, train$E, y ~ G*E, family=binomial)
</code></pre>

<pre><code>## Converged in 6 iterations
</code></pre>

<pre><code class="r">summary(fit_default)
</code></pre>

<pre><code>## $fit_main
## 
## Call:
## stats::glm(formula = formula, family = family, data = data, model = FALSE, 
##     y = FALSE)
## 
## Deviance Residuals: 
##     Min       1Q   Median       3Q      Max  
## -2.2401  -0.5533  -0.1316   0.4363   3.0637  
## 
## Coefficients: (-7 not defined because of singularities)
##             Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept)  -0.9892     0.1065  -9.293  &lt; 2e-16 ***
## G             2.0863     0.6747   3.092 0.001989 ** 
## E             3.2313     0.2148  15.042  &lt; 2e-16 ***
## G:E           4.3265     1.1960   3.617 0.000298 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 1351.50  on 999  degrees of freedom
## Residual deviance:  692.69  on 989  degrees of freedom
## AIC: 714.69
## 
## Number of Fisher Scoring iterations: 6
## 
## 
## $fit_genes
## 
## Call:
## stats::glm(formula = formula_b, family = family, data = data, 
##     model = FALSE, y = FALSE)
## 
## Deviance Residuals: 
##     Min       1Q   Median       3Q      Max  
## -2.3976  -0.5512  -0.1208   0.4385   3.0640  
## 
## Coefficients: (-5 not defined because of singularities)
##       Estimate Std. Error z value Pr(&gt;|z|)    
## g1     0.12059    0.07512   1.605  0.10842    
## g2     0.08314    0.06927   1.200  0.23005    
## g3    -0.28041    0.05142  -5.454 4.94e-08 ***
## g4     0.02208    0.05359   0.412  0.68030    
## g1_g3  0.06894    0.12845   0.537  0.59146    
## g2_g3  0.42484    0.15154   2.804  0.00505 ** 
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 1351.50  on 1000  degrees of freedom
## Residual deviance:  692.69  on  989  degrees of freedom
## AIC: 714.69
## 
## Number of Fisher Scoring iterations: 6
## 
## 
## $fit_env
## 
## Call:
## stats::glm(formula = formula_c, family = family, data = data, 
##     model = FALSE, y = FALSE)
## 
## Deviance Residuals: 
##     Min       1Q   Median       3Q      Max  
## -2.3965  -0.5513  -0.1277   0.4389   3.0637  
## 
## Coefficients: (-8 not defined because of singularities)
##    Estimate Std. Error z value Pr(&gt;|z|)    
## e1 -0.43085    0.02905 -14.832   &lt;2e-16 ***
## e2  0.35751    0.02613  13.681   &lt;2e-16 ***
## e3  0.21164    0.02156   9.814   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 1351.50  on 1000  degrees of freedom
## Residual deviance:  692.69  on  989  degrees of freedom
## AIC: 714.69
## 
## Number of Fisher Scoring iterations: 6
</code></pre>

<p>We are a little off, especially with regards to the weights of the genetic variants. This is because there is substantial loss of information with binary outcomes. To assess the quality of the fit, we are going to do a 5-Folds cross-validation.</p>

<pre><code class="r">cv_5folds_bin = LEGIT_cv(train$data, train$G, train$E, y ~ G*E, cv_iter=1, cv_folds=5, classification=TRUE, family=binomial, seed=777)
pROC::plot.roc(cv_5folds_bin$roc_curve[[1]])
</code></pre>

<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfgAAAH4CAMAAACR9g9NAAAAjVBMVEUAAAAAADoAAGYAOjoAOpAAZrY6AAA6ADo6AGY6OmY6OpA6ZmY6ZrY6kNtmAABmADpmAGZmOgBmOpBmZgBmZjpmZmZmtv+QOgCQOjqQOmaQZgCQkDqQtpCQ27aQ29uQ2/+pqam2ZgC2Zjq225C2/7a2///bkDrb25Db/7bb////tmb/25D//7b//9v///880fpBAAAACXBIWXMAAAsSAAALEgHS3X78AAAPgUlEQVR4nO3dDXfiTBnGcbZ267paWleFVR/tVp/yWN6+/8eTpLCEkkIm98zcL9f1P3qw58hA+9tJJiFpJ1sG2UT7DTCdCA8a4UEjPGiEB43woBEeNMKDRnjQCA8a4UEjPGiEB43woBEeNMKDRnjQCA8a4UEjPGiEB43woBEeNMKDRnjQCA8a4UEjPGiEB00CP2GG+1EQXvBcVrgfV3QIH7Mf13QIH7IfV3UIH7Fm/054vNp1HeHhelvPEx6t/XGcGH511x4W3ryMeC6r3+H4XQq/mc/ax+Xta/JzWf1+nreRwq8fX04eU57Lqnc8X8cZj1TnPK14H7+ech/vpe75ea7qcTr5XIbwrkv4OO7UPRt8Z3F3fCmWpzwfxL4b88pLSt6u4LmwyTQv9P7zd8IXL/tkHdPZdReEL5iOcV/n19sQPkMqlin1XGclPnM33X935wfyBr7hwilN3+T6rq8Tz/jN/H70c31mF/iDeq+rlG/q1w9Po5/rIvsz+nL919NyH385t9w/++A6asJfyCX0uz66fp7wvXmd3md9eN8E4ff53Ydf6uP7ZQjfFo+87cJ9UvDwwahPunR/HDZ8tBl+2sX7IqHhw5K3XXSHhQ880fdddkeDD7mE6+2KOxQ8innTNXckeATvQ1fdceCB1Ie4w8DT/V048NrvoF5D3EHgOd/PQoBHWtUNdQ8Pj3Hw1mmge3B4OPbB7rHhwdC3Ce5x4eHmetNw96jweNv4pgT3gPAgJ+N7SnGPBw/LnuYeEV7lZfVLcyd8lBLdCR+kVHfCxyjZPRw813UDiwQPu6Af4R4JHpV9lHss+MovaKRR7mHgUWf7WPco8LCb+bHuceArvpilxroT3nej3YPAg27mBe4x4Omennt43GWdyN09PN1H5h++xqtYTObuHR51tovdncPTfXSe4WH37hncPcPjLusyuDuGh1XP4u4XHne6Z3H3DF90eMNlcXcLz/kuzCU88LIul7tLeLpnyBs88Ln5pmzu3uDpnitX8NDo26zuruDpnjFH8ODsed19wWce0Fd53R3Bg0/4zO6u4POO56vc7n7gsSd8dnc38Ngr+vzuXuCh2Uu4y+FXd5PPi8nk5mXEc4dH99xJ4Tffn7aLzzv/r6/Jzx0e9IQv4i6GXz++bBf3b4+pzx0e3bPnZcbnG8tZhdyz7OPvi+/jceFLuftY1ePu4ou5O4BH/gi+nHs2+M7iLu9f8QRmL+luf8bDqpd1tw4PPN3LumdZ1TeVWdVzO18s8XH8fNY+Lm8LHMfjqhd3z3LmrvuY8twrAc/28u6GZzzdiybex6+nZfbxyOwV3O2u6uleNsLbq4a7WXjgLX0Vd8Pwkmd7ro474a1Vyd0qPOyWvpa7XXjBkx1XzZ3wpqrnbhQedEtf0d0mPN3LZxRe8LJuq+puDx72Eru67ubg6V4pe/CCV3RcbXdr8JizXcHdGDzoZl7B3Ry84PX8puBOeANpuBNePxV3wqun40547ZTcCa+cljvhdVNzJ7xqeu624NFO3yi6W4MXvJy/NN1NwYNNeFV3Y/CCV3OXrjvhtVJ2J7xS2u6G4KE+klV3twMPdcmVvrsleMFLOcuAO+EVsuBuBJ7b+eqZgKd7/YzAC17GWUbcCV85K+6Er5sZd8JXzY67CXiYlZ0hdwvwdNfIBLzgRRxlyp3w1bLlTvhaGXMnfKWsuRO+TubcCV8le+6Er5FBdwPw8Q/jLbqbgBe8hodMuhO+eDbdCV86o+6EL5xVd3342Gs7s+4W4AUvYT277oQvmWF3whfMsjvhy2XanfDFsu1O+FIZdyd8oay7y+FXd7PNfDI5//Px0PDm3cXwm/ls+zzb+X89kx8kGvP8jX13Mfz68WXz/al9TH5u+3+iu07iTf1uui/vt9vl5xHP3caE9+CeYXH33N7kfO4OC+/CXX1VHw/eh7s2fLy1nRP3bPCdxd3k0IBXDwfvxV13xodj9+OuCk93xTKcuWs36zdnh/FD4K+O7itH7lnO3DUtz8/ZXntuuAnvyT3HmbuTx4Sho8G7clec8cHYnbnL9/Hr6ch9fDB4Z+6Kq/pY8N7cCZ8nd+6Ez5I/d8LnyKF7Gvx62vPp6+ihw8B7dE+d8cvJ5NNTpqGjwLt0H7Gpby6tnOUYOgi8T/dU+NVdM+N7TtONGDoGvFP31H18z1XUo4cOAe/VPRG+nenD5vu1oWOcqXfrngK/Pznbe/NE8tAxPqLx6z5mxmcZOgK7Z3etEzic79qlbOoff/3wk7jUoemundqMFwxsI9/uhB+bc3et43j38N7dR5y567tNLnlo7/Du3cds6pfyxZ33Rb1/d6UZT3f1dPbxvuEjuCut6l3Dh3BXOoHjGT6GO2d8akHclT6W9QsfxV3pY1m38GHclT6W9Qofx537+JQCuXNVn1Akd8744YVyJ/zgYrknwi9uXxcDb6eIBh/MPXFV//C0+8/qi3Qf7/DDuWjuyYdzuzmfA37QAIYK5566qZ98elqKN/X+Jnw8d5XFHd0NRPjrRXRPhF9mOYHjDD6ke+oVOAN375eH9gUf013lQxpX8EHdEzf1z/c5hvYEH9U9dVOPto8P685V/cXiuhP+UoHdE+E388nt/x4G/r4z9/CR3dPgN/P71dfXnt9QnjS0lzO2od2TD+d28MKrbL389pvY7mNm/EI0452wR3dP38cPvrr6Q/hhz1Yuunv9Vb0P+PDuhO8tvnsSfPNby1d3Q0/cOYYHcE+Cf75v/+jUYuBvRnALj+CedkPFy7a53k52OGcfHsI9Fb45lJNdbGkeHsM9bVM/2x3I7x4km3rzh/Eg7omLu90x/PC/S9M3tPnTdijutQ/njLPjuBP+JBx3wncDcq8Mb3sPj+SeCb73CK8Xfth4KkG5i+F//kak8xO5Z8+1vaTHcpfP+PX05mXYjKe7pTJs6tfT29+GwQ99UwqhuefZx6/u+j6x8wQP515zVW8YHs+d8E2A7tngO5/VTg69H8sqPKI7ZzyoO+FB3eXwzV+p6b+D1gc8qLsYvrkIr6nnvioX8Kju8lO2+0Vdz4V4HuBh3cFnPK57jnP1fvfxwO7Qq3pkd2R4aHdgeGx3XHhwd1h4dHdUeHh3UHi6Y8LTHROe7tua8GausaV7U014wVgZo3sbHDzd30KDp/s+MHi6H8KCp/vPoODpfgwJnu6dgODp3g0Hnu4nwcDT/TQUeLq/CwSe7u/DgKf7WRDwdD8PAZ7uPQHA072v+PB07y08PN37iw5P9w8KDk/3j4oNT/cPCw1P94+LDE/3CwWGp/ul4sLT/WJh4el+uajwdL9SNfi6t87R/VoV4QVDpUb3q4WEp/v1IsLTfUAB4ek+pHjwdB9UOHi6DysaPN0HFgye7kOLBU/3wYWCp/vwIsHTPaFA8HRPKQ483ZMKA0/3tKLA0z2xIPB0Ty0GPN2TCwFP9/QiwNN9RAHg6T4m//B0H5V7eLqPyzs83UfmHJ7uYxPDr+4mn5622/Xj2d8RrwBP99FJ4Tfz2e6/9zrwdB+fFP4N/PmzBjzdBeWY8bsWv/tSHZ7uksT7+PX0vnlY3NSGp7sot6t6usvyCk93YbngO4u7yaHTkbLC012azxlPd3Eu4ekuL8eZu6bzRX05eLpnKNNx/HZ5+3r5ufng6Z6jPGfuap6rp3uW3M14uucpw5m7qvt4umfK2aqe7rnyBU/3bLmCp3u+PMHTPWOO4OmeMz/wdM+aG3i6580LPN0z5wSe7rnzAU/37LmAp3v+PMDTvUAO4OleIvvwdC+SeXi6l8k6PN0LZRye7qWyDU/3YpmGp3u5LMPTvWCG4eleMrvwdC+aWXi6l80qPN0LZxSe7qWzCU/34pmEp3v5LMLTvUIG4eleI3vwdK+SOXi618kaPN0rZQye7rWyBU/3apmCp3u9LMHTvWKG4OleMzvwdK+aGXi6180KPN0rZwSe7rWzAU/36pmAp3v9LMDTXSED8HTXSB+e7iqpw9NdJ214uiulDE93rXTh6a6WKjzd9dKEp7tiivB010wPnu6qqcHTXTcteLorpwRPd+104Omungo83fXTgKe7gRTg6W6h+vB0N5EYfnWX9mfE6W4jKfxmPmsfl7evl5+7h6e7kaTw68eXk8cPn/sGT3cr1Z3xdDeTeB+/ng7fx9PdTjVX9XQ3VEV4ulsqF3xncTc5dDqS5N8Qy161Gc/5bqta8HQ3VqUzd3S3Vp3jeLqbq8qZO7rbq8aMp7vBKpy5o7vFyq/q6W6y4vB0t1lpeLobrTA83a1WFp7uZisKT3e7lYSnu+EKwv+YMMOVgy83VN7BzL4xzcEIDzoY4UEHIzzoYIQHHYzwoIMRHnQwwoMOxovhQSM8aIQHjfCgER40woNGeNAIDxrhQcsCv/qyv79uOjm/0Sqt4xCru767d8aNtZlPPj3lemPH28oyDNbci5xtsJSffw745V6o+XksPouGOg6xfnjaLkT/jDpv53nWd+/fyMG224XQ6vS7XP1e9G/yOFjSzz8D/POnX95mfHND7WHyj+w4xOrra98duqPGko3zbrDdW/vDn2Xwx8GWjdOzaLTTb3Pwzz/npr61ehD98z0OIZ/xx7FWX/8h3dR3vrfN938LN/WnP6hsP7Kkn39O+GZrKvwuOkOIFwzHsVZ3s/ankueNLe6l+/iTH9Rmfp9psKSfv9kZ3+z5lqLV3cipMGAwKXz3/aynMncLMz7rPl6++ejs/L6J4Y+DLdqL1kVanR9UszGSZWAf32y0xKv6wxDyGd95O8/iTf3J9yad8Z3vUuzeGSzp558NvvlvtuP4ZrDlRLogO461+1/CcwKdwbIdx+8Ge9t8CI8Rut9m1eN45jDCg0Z40AgPGuFBIzxohAeN8KARHjTCg0Z40AgPGuFBg4Nvfv3+B5/Urb68bOaTPx4+0d59Kby6wHJo8OvpbHvhWr530oQPU0u5fnhaP/6z/bh//0n97uHT0+rLr7uHX94+2N5/efMX8TUcNkOD38zfZvt6evvaXODz3F600lxZsbz9rd22txv845fL++1SeFWcydDgt/sre5pN/ub7U3Ml3vpxvy9fHeC7X66/vf5HeBOOyfDgt+31fO21l8+z9i8t7bbq7cb8CN/5cvP9X98Cbunh4NtbVxrxHXwz49/usbkw47eLv0Xc0sPBt6v6ZnE3/dxewb3bx+8emp366st/u/v4w5fSW9ushgbfHsc3+/iHvx5W9T8fjtP98OVmfvOy+XvIQzo4+H3D769Y/anoG9GK8FdaSG+sNxoqPHyEB43woBEeNMKDRnjQCA8a4UEjPGj/ByIid9EuoTifAAAAAElFTkSuQmCC" alt="plot of chunk unnamed-chunk-12"/>
Although the weights of the genetic variants are a bit off, the model predictive power is good.</p>

</body>

</html>
